src/graphql/graphql-api.php

```diff
@@ -0,0 +1,245 @@
+				$inner_blocks_copy = $inner_block['innerBlocks'];
+				unset( $inner_block['innerBlocks'] );
+				$inner_block['parentId'] = $inner_block['parentId'] ?? $filter_options['parentId'];
+				array_push( $flattened_blocks, $inner_block );
+				$flattened_blocks = self::flatten_inner_blocks( $inner_blocks_copy, $filter_options, $flattened_blocks );
+			}
+		}
+
+		return $flattened_blocks;
+	}
+
+	/**
+	 * Register types and fields graphql integration.
+	 *
+	 * @return void
+	 */
+	public static function register_types() {
+		// Register the type corresponding to the attributes of each individual block.
+		register_graphql_object_type(
+			'BlockDataAttribute',
+			[
+				'description' => 'Block data attribute',
+				'fields'      => [
+					'name'  => [
+						'type'        => 'String',
+						'description' => 'Block data attribute name',
+					],
+					'value' => [
+						'type'        => 'String',
+						'description' => 'Block data attribute value',
+					],
+				],
+			],
+		);
+
+		// Register the type corresponding to the individual inner block, with the above attribute.
+		register_graphql_type(
+			'InnerBlockData',
+			[
+				'description' => 'Block data',
+				'fields'      => [
+					'id'         => [
+						'type'        => 'String',
+						'description' => 'ID of the block',
+					],
+					'parentId'   => [
+						'type'        => 'String',
+						'description' => 'ID of the parent for this inner block, if it is an inner block. This will match the ID of the block',
+					],
+					'name'       => [
+						'type'        => 'String',
+						'description' => 'Block name',
+					],
+					'attributes' => [
+						'type'        => [
+							'list_of' => 'BlockDataAttribute',
+						],
+						'description' => 'Block data attributes',
+					],
+				],
+			],
+		);
+
+		// Register the type corresponding to the individual block, with the above attribute.
+		register_graphql_type(
+			'BlockData',
+			[
+				'description' => 'Block data',
+				'fields'      => [
+					'id'          => [
+						'type'        => 'String',
+						'description' => 'ID of the block',
+					],
+					'name'        => [
+						'type'        => 'String',
+						'description' => 'Block name',
+					],
+					'attributes'  => [
+						'type'        => [
+							'list_of' => 'BlockDataAttribute',
+						],
+						'description' => 'Block data attributes',
+					],
+					'innerBlocks' => [
+						'type'        => [ 'list_of' => 'InnerBlockData' ],
+						'description' => 'Flattened list of inner blocks of this block',
+					],
+				],
+			],
+		);
+
+		// Register the type corresponding to the list of individual blocks, with each item being the above type.
+		register_graphql_type(
+			'BlocksData',
+			[
+				'description' => 'Data for all the blocks',
+				'fields'      => [
+					'blocks'   => [
+						'type'        => [ 'list_of' => 'BlockData' ],
+						'description' => 'List of blocks data',
+					],
+					'warnings' => [
```

Comment:
Warning / error data typically doesn't ship in the `data` payload, but instead [goes to `graphql_debug()`](https://www.wpgraphql.com/docs/debugging#graphql_debug) where it can be conditionally displayed in the `errors` payload. Better control for the site maintainer.

Reply:
I'm still working on this, ran into some problems that are of my own making.